src/main/java/org/pentaho/di/trans/steps/annotation/BaseAnnotationMeta.java

```diff
@@ -0,0 +1,455 @@
+
+        List<ColumnMapping> columnMappings = new ArrayList<ColumnMapping>();
+        long nrColumnMappings = rep.getStepAttributeString( id_step, "CM_COUNT_" + i ) != null
+            ? Long.valueOf( rep.getStepAttributeString( id_step, "CM_COUNT_" + i ) )
+            : 0;
+        for ( int j = 0; j < nrColumnMappings; j++ ) {
+
+          ColumnMapping columnMapping = new ColumnMapping();
+
+          columnMapping.setName( rep.getStepAttributeString( id_step, i, "CM_NAME_" + j ) );
+          columnMapping.setColumnName( rep.getStepAttributeString( id_step, i, "CM_COLUMN_NAME_" + j ) );
+          String dataType = rep.getStepAttributeString( id_step, i, "CM_DATA_TYPE_" + j );
+          if ( StringUtils.isNotBlank( dataType ) ) {
+            columnMapping.setColumnDataType( DataType.valueOf( dataType ) );
+          }
+
+          columnMappings.add( columnMapping );
+        }
+
+        dataProvider.setColumnMappings( columnMappings );
+        dataProviders.add( dataProvider );
+      }
+      modelAnnotationGroup.setDataProviders( dataProviders );
+
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString(
+          PKG, "ModelAnnotationMeta.Exception.UnexpectedErrorReadingStepInfoFromRepository" ), e );
+    }
+
+    setModelAnnotations( modelAnnotationGroup );
+
+    // This may override the loaded model annotation group
+    if ( StringUtils.isNotBlank( getModelAnnotationCategory() ) ) {
+      readDataFromMetaStore( metaStore );
+    }
+  }
+
+  public void readDataFromMetaStore( IMetaStore metaStore ) {
+    try {
+
+      if ( metaStore == null ) {
+        return;
+      }
+
+      ModelAnnotationManager manager = getModelAnnotationManager( getModelAnnotations() );
+      if ( manager.containsGroup( getModelAnnotationCategory(), metaStore ) ) {
+        setModelAnnotations( manager.readGroup( getModelAnnotationCategory(), metaStore ) );
+      }
+    } catch ( Exception e ) {
+      logError( e.getMessage() );
+    }
+  }
+
+  public void saveToMetaStore( IMetaStore metaStore ) throws Exception {
+    saveToMetaStore( metaStore, getModelAnnotations() );
+  }
+
+  public void saveToMetaStore( IMetaStore metaStore, ModelAnnotationGroup modelAnnotations ) throws Exception {
+    if ( metaStore == null ) {
+      return;
+    }
+
+    ModelAnnotationManager manager = getModelAnnotationManager( modelAnnotations );
+    if ( modelAnnotations != null && StringUtils.isBlank( modelAnnotations.getName() ) ) {
+      modelAnnotations.setName( this.getName() );
+    }
+    checkValidName( modelAnnotations.getName() );
+    manager.createGroup( modelAnnotations, metaStore );
+  }
+
+  public void checkValidName( final String name ) throws KettleException {
+    // potentially problematic characters in filesystem/repository metastores
+    char[] fileSystemReservedChars = new char[] { '\\', '/', ':', '*', '?', '"', '<', '>', '|', '\t', '\r', '\n' };
```

Comment:
maybe, but since this is moving that line from ModelAnnotationMeta to this new class, i don't think this case is the place to do it